### PR TITLE
[expo-updates] Add assets:verify command to CLI

### PR DESCRIPTION
**Pull request by douglowder**
# Why

Command for use by applications using `expo-updates`, to verify that update bundles that use [asset selection/exclusion](https://docs.expo.dev/eas-update/asset-selection/) contain all assets not found in the embedded manifest for a native build.

# How

The new command takes a platform and the paths to the various files it needs from an exported bundle and from a build. Using the asset dump file, it constructs the set of all asset hashes resolved by Metro, and verifies that each of those assets are found in either the build or the export bundle.  Warning messages are printed if there are any missing assets.

```sh
$ npx expo-updates assets:verify --help

Description
Verify that all static files in an exported bundle are in either the export or an embedded bundle

Usage
  $ npx expo-updates assets:verify <dir>

  Options
  <dir>                                  Directory of the Expo project. Default: Current working directory
  -a, --asset-map-path <path>            Path to the `assetmap.json` in an export produced by the command `npx expo export --dump-assetmap`
  -e, --exported-manifest-path <path>    Path to the `metadata.json` in an export produced by the command `npx expo export --dump-assetmap`
  -b, --build-manifest-path <path>       Path to the `app.manifest` file created by expo-updates in an Expo application build (either ios or android)
  -p, --platform <platform>              Options: ["android","ios"]
  -h, --help                             Usage info
  

```

# Test Plan

- Tested against a local application built with asset selection
- New unit tests

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/bin/cli.ts
Comment (byCedric) [inline packages/@expo/cli/bin/cli.ts:30]: Right now, `verify-native-assets` will appear in `npx expo --help` output. Since this looks like an internal command, you'll also need to explicitly [pull this command out here](https://github.com/expo/expo/blob/c8b3f7bcf9f50980478e41e831f7b70550d8a2b4/packages/%40expo/cli/bin/cli.ts#L90-L97) to hide it.

This command name also seems to deviate a bit from the format we use in other commands. E.g. `npx expo asset:validate` etc.

File: packages/@expo/cli/src/verifyNativeAssets/resolveOptions.ts
Comment (byCedric) [inline packages/@expo/cli/src/verifyNativeAssets/resolveOptions.ts:16]: I think `platform` needs an assertion here. What if `--platform web` is used?

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (byCedric) [inline packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts:76]: Even though we type it to only `ios | android`, it might make sense to explicitly support `ios` instead of defaulting to it. That way, if ]platform is ever changed to something like `macos`]
(https://github.com/expo/expo/pull/26292), it is not silently inheriting `ios`.

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (byCedric) [inline packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts:162]: Same here

**Review by byCedric — DISMISSED** [review]
Awesome, nice addition to clarify how assets work in Expo Updates. Left a few notes.

Also had a question if this belongs in `npx expo`, or if it should exist in `eas-cli`, `npx expo-updates` (specific updates cli), or even doctor. It seems a bit out of place for specific update commands in `npx expo`, as we [don't have any domain logic in this CLI related to updates.](https://github.com/search?q=repo%3Aexpo%2Fexpo+path%3A%2F%5Epackages%5C%2F%40expo%5C%2Fcli%5C%2F%2F+expo-updates&type=code). 

Maybe this fits `eas-cli` better, as that's the CLI with the most domain logic related to updates. But, maybe it's a specialised command for `expo-updates` itself, [like codesigning](https://github.com/expo/expo/blob/main/packages/expo-updates/cli/cli.ts).

File: packages/@expo/cli/src/verifyNativeAssets/index.ts
Comment (byCedric) [inline packages/@expo/cli/src/verifyNativeAssets/index.ts:63]: Is it correct that the command always exits with status code `0`, even when there are missing assets?

**Comment by douglowder** [comment]
> maybe it's a specialised command for expo-updates itself, like codesigning.

@byCedric this is it exactly!

File: packages/@expo/cli/src/verifyNativeAssets/index.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/index.ts:63]: It will throw under some error conditions, but missing assets will still exit 0. It's not a bad idea to have missing assets cause an exit 1 instead.

Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/index.ts]: Modified the code to throw a CommandError if assets are missing, this seems semantically correct...

**Comment by wschurman** [comment]
> > maybe it's a specialised command for expo-updates itself, like codesigning.
> 
> @byCedric this is it exactly!

So we do want to move this to the expo-updates CLI (that's where the codesigning commands live)?

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (wschurman) [inline packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts:80]: this overwrites what is being set in the switch?

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (wschurman) [inline packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts:74]: > Path to an EAS or local build containing an expo-updates embedded manifest (app.manifest)

At first I was under the impression that one could pass in the final build apk, aab, or ipa, when in reality they need the build artifacts/intermediates. I'm not sure how reliable this method (using build intermediates) will be for determining compatibility for a few reasons:
- artifacts are deleted/cleaned/removed often, and sometimes even automatically. the final build is often the only output that is kept.
- locally, it is overwritten for each build and often stale artifacts/intermediates remain in the unclean'ed directory from prior builds.
- users won't have these for an EAS build.

So, I guess my question is about proposed usage of this. Are we requiring users to retain copies of these intermediates? Or, should we provide a way to decompile the binary to extract the app.manifest? Or, should we just record this info when making an EAS build and provide that info in a raw form into this command?

File: packages/@expo/cli/src/verifyNativeAssets/index.ts
Comment (wschurman) [inline packages/@expo/cli/src/verifyNativeAssets/index.ts:59]: It might be preferable to print out the assets that are missing here rather than in `getMissingAssets`. This way if that function is ever needed somewhere that doesn't desire them to be warn'ed on the console, modification isn't needed. Also slightly prefer the output to read:

```
7 assets not found in either embedded manifest or in exported bundle:
- asset1
- asset2
- ...
```

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (wschurman) [inline packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts:132]:
```suggestion
  return new Set(assetDump.keys());
```

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (wschurman) [inline packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts:171]:
```suggestion
  return new Set(assets.map((asset) => asset.path.substring(7, asset.path.length)));
```

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (wschurman) [inline packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts:169]: nit: add a comment indicating what the "7" is here

File: packages/@expo/cli/src/verifyNativeAssets/index.ts
Comment (wschurman) [inline packages/@expo/cli/src/verifyNativeAssets/index.ts:62]:
```suggestion
      console.info(`All resolved assets found in either embedded manifest or in exported bundle.`);
```
since everything is okay, no need to warn?

**Review by wschurman — CHANGES_REQUESTED** [review]

File: packages/@expo/cli/src/verifyNativeAssets/index.ts
Comment (byCedric) [reply packages/@expo/cli/src/verifyNativeAssets/index.ts:62]: Yeah, I think it's mostly relevant if we add `--json` output support. In that case, we don't want to mix up JSON with the actual logs.

But we aren't doing that in `npx expo` anyways (splitting into warn and log) 🤷

File: packages/@expo/cli/src/verifyNativeAssets/resolveOptions.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/resolveOptions.ts]: Fixed

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: Since platform is now required, it is probably safe to leave this in

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: Platform is now a required parameter. I'll take a look to see if this code can be adjusted.

File: packages/@expo/cli/bin/cli.ts
Comment (douglowder) [reply packages/@expo/cli/bin/cli.ts]: Command name is now `assets:verify`

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: @wschurman the apk or ipa would be fine, since they would presumably have the `app.manifest` file. They would have to be unpacked first.

Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: There is already an `apktool` to unpack Android apks, and I think ipas are just zip or gzip files.

Comment (wschurman) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: Yep, I think we just need to be deliberate about what the input to this function should be. IMO it makes the most sense to pass the app.manifest file path explicitly so that this command isn't responsible for digging in IPA/AAR/APK/ZIP files or build intermediates. It also leaves the window open for supplying the app.manifest directly from EAS servers, which would potentially be a good UX at update time.

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: Good catch, this has been fixed.

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: @wschurman now that I've moved this to the expo-updates CLI, it might be good to be able to pass the app.manifest directly with a different command option. I'll try adding that next.

File: packages/@expo/cli/src/verifyNativeAssets/index.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/index.ts]: Done

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: Done

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: Done

File: packages/@expo/cli/src/verifyNativeAssets/index.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/index.ts]: Done

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (douglowder) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: Done

File: packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts
Comment (wschurman) [reply packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts]: Sounds good. My take is that this command should be as single-responsibility as possible and just take the two file paths as arguments (the assetmap and the app.manifest). The single responsibility is determining what the missing assets are as opposed to that plus knowing the structure of builds/dist folder. My guess is that this will most often be run by us from eas-cli or something, and that CLI knows exactly where these two files live (it does the export and also does the build and can fetch build metadata/app.manifest for a build from the servers).

**Review by wschurman — APPROVED** [review]
Approving to unblock. Before landing, I think it's still best to change the args to just the file paths as discussed inline but no rush and I'll leave it up to you.

**Review by wschurman — APPROVED** [review]

**Pull request merged**
